File: packages/sdk-ruby/lib/melaya/templates.rb
# frozen_string_literal: true

module Melaya
  # Templates API — create, manage, share, and assign pipeline templates.
  #
  # Templates bundle a pipeline definition into a reusable, shareable artifact.
  # Visibility levels: private → team → community → assigned.
  #
  # Maps to:
  #   /api/v1/private/user-templates/*    — CRUD + share + assignments
  #   /api/v1/private/templates/*         — global/validated lists
  #
  # @example
  #   templates = melaya.templates.list
  #   t = melaya.templates.save(name: "My report", payload: { steps: [] })
  #   melaya.templates.share(t["id"], "team")
  #   melaya.templates.delete(t["id"])
  class TemplatesAPI
    def initialize(http)
      @http = http
    end

    # GET /api/v1/private/user-templates
    # List all templates visible to the caller (own + team + community + assigned).
    def list
      @http.get("/api/v1/private/user-templates")
    end

    # GET /api/v1/private/templates/global
    # List all community-visibility (global) templates.
    def list_global
      @http.get("/api/v1/private/templates/global")
    end

    # GET /api/v1/private/templates/validated
    # List IDs of all validated (platform-approved) templates.
    def list_validated
      @http.get("/api/v1/private/templates/validated")
    end

    # POST /api/v1/private/user-templates
    # Create a new private user template.
    # @param name [String]
    # @param payload [Hash]
    # @param description [String, nil]
    # @param category [String, nil]
    def save(name:, payload:, description: nil, category: nil)
      body = compact(
        "name"        => name,
        "payload"     => payload,
        "description" => description,
        "category"    => category
      )
      @http.post("/api/v1/private/user-templates", body)
    end

    # PATCH /api/v1/private/user-templates/:id
    # Update name/description/category/payload of a private user template.
    # @param template_id [String]
    # @param body [Hash]
    def update(template_id, body = {})
      @http.patch("/api/v1/private/user-templates/#{enc(template_id)}", body)
    end

    # POST /api/v1/private/user-templates/:sourceId/duplicate
    # Duplicate a readable template into the caller's private library.
    # @param template_id [String] the source template to copy
    # @param new_name [String, nil]
    def duplicate(template_id, new_name: nil)
      body = new_name ? { "newName" => new_name } : nil
      @http.post("/api/v1/private/user-templates/#{enc(template_id)}/duplicate", body)
    end

    # DELETE /api/v1/private/user-templates/:id
    # Delete (or soft-demote if shared) a template.
    # @param template_id [String]
    def delete(template_id)
      @http.delete("/api/v1/private/user-templates/#{enc(template_id)}")
    end

    # PUT /api/v1/private/user-templates/:id/visibility
    # Change the visibility of a template.
    # @param template_id [String]
    # @param visibility [String] "private", "team", "community", or "assigned"
    def share(template_id, visibility)
      @http.put("/api/v1/private/user-templates/#{enc(template_id)}/visibility",
        "visibility" => visibility)
    end

    # GET /api/v1/private/user-templates/share-targets
    # List projects the caller is a member of (for the share target picker UI).
    def share_targets
      @http.get("/api/v1/private/user-templates/share-targets")
    end

    # ── Assignments ────────────────────────────────────────────────────────────

    # GET /api/v1/private/user-templates/:templateId/assignments
    # List all assignments (users / projects) for a template.
    # @param template_id [String]
    def list_assignments(template_id)
      @http.get("/api/v1/private/user-templates/#{enc(template_id)}/assignments")
    end

    # POST /api/v1/private/user-templates/:templateId/assignments
    # Assign a template to a user or project.
    # Provide exactly one of +user_id:+ or +project_id:+ (both UUIDs);
    # the server rejects requests carrying both or neither.
    # @param template_id [String]
    # @param user_id [String, nil] target user UUID
    # @param project_id [String, nil] target project UUID
    def assign(template_id, user_id: nil, project_id: nil)
      @http.post("/api/v1/private/user-templates/#{enc(template_id)}/assignments",
        assignment_target(user_id, project_id))
    end

    # DELETE /api/v1/private/user-templates/:templateId/assignments
    # Remove an assignment from a template.
    # Provide exactly one of +user_id:+ or +project_id:+ (both UUIDs).
    # The target is sent as query params — the server ignores DELETE
    # request bodies.
    # @param template_id [String]
    # @param user_id [String, nil] target user UUID
    # @param project_id [String, nil] target project UUID
    def unassign(template_id, user_id: nil, project_id: nil)
      @http.delete("/api/v1/private/user-templates/#{enc(template_id)}/assignments",
        assignment_target(user_id, project_id))
    end

    private

    # Exactly one of user_id / project_id must be given.
    def assignment_target(user_id, project_id)
      if user_id.nil? == project_id.nil?
        raise ArgumentError, "Melaya: provide exactly one of user_id: or project_id:"
      end
      user_id ? { "userId" => user_id } : { "projectId" => project_id }
    end

    def enc(s)
      URI.encode_www_form_component(s.to_s)
    end

    def compact(hash)
      hash.reject { |_, v| v.nil? }
    end
  end
end

File: packages/sdk-rust/src/templates.rs
use serde_json::{json, Value};
use std::collections::HashMap;

use crate::client::HttpClient;
use crate::error::Result;

/// Templates API — create, manage, and share pipeline templates.
#[derive(Clone)]
pub struct TemplatesAPI {
    http: HttpClient,
}

impl TemplatesAPI {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    /// List all templates visible to the caller — own + team + community + assigned.
    pub async fn list(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http.get("/api/v1/private/user-templates", &q).await
    }

    /// List all community-visibility (global) templates.
    pub async fn list_global(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http.get("/api/v1/private/templates/global", &q).await
    }

    /// List IDs of all validated (platform-approved) templates.
    pub async fn list_validated(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get("/api/v1/private/templates/validated", &q)
            .await
    }

    /// Create a new private user template.
    pub async fn save(&self, body: &Value) -> Result<Value> {
        self.http.post("/api/v1/private/user-templates", body).await
    }

    /// Update name/description/category/payload of a private user template.
    pub async fn update(&self, template_id: &str, body: &Value) -> Result<Value> {
        self.http
            .patch(
                &format!("/api/v1/private/user-templates/{template_id}"),
                body,
            )
            .await
    }

    /// Duplicate a readable template into the caller's private library.
    pub async fn duplicate(&self, template_id: &str, new_name: Option<&str>) -> Result<Value> {
        let mut body = json!({});
        if let Some(n) = new_name {
            body["newName"] = json!(n);
        }
        self.http
            .post(
                &format!("/api/v1/private/user-templates/{template_id}/duplicate"),
                &body,
            )
            .await
    }

    /// Delete (or soft-demote if shared) a template.
    pub async fn delete(&self, template_id: &str) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .delete(&format!("/api/v1/private/user-templates/{template_id}"), &q)
            .await
    }

    /// Change the visibility of a template.
    /// `visibility` is one of: `"private"`, `"team"`, `"community"`, `"assigned"`.
    pub async fn share(&self, template_id: &str, visibility: &str) -> Result<Value> {
        self.http
            .put(
                &format!("/api/v1/private/user-templates/{template_id}/visibility"),
                &json!({ "visibility": visibility }),
            )
            .await
    }

    // ── Assignments ──────────────────────────────────────────────────────────

    /// List all assignments (users / projects) for a template.
    pub async fn list_assignments(&self, template_id: &str) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get(
                &format!("/api/v1/private/user-templates/{template_id}/assignments"),
                &q,
            )
            .await
    }

    /// Assign a template to a user or project.
    ///
    /// Provide exactly one of `user_id` / `project_id` (both UUIDs) — the
    /// server rejects requests carrying both or neither. The target is sent
    /// in the JSON body as `{ "userId": ... }` or `{ "projectId": ... }`.
    pub async fn assign(
        &self,
        template_id: &str,
        user_id: Option<&str>,
        project_id: Option<&str>,
    ) -> Result<Value> {
        let mut body = json!({});
        if let Some(u) = user_id {
            body["userId"] = json!(u);
        }
        if let Some(p) = project_id {
            body["projectId"] = json!(p);
        }
        self.http
            .post(
                &format!("/api/v1/private/user-templates/{template_id}/assignments"),
                &body,
            )
            .await
    }

    /// Remove an assignment from a template.
    ///
    /// Provide exactly one of `user_id` / `project_id` (both UUIDs). The
    /// target travels as query params — the REST bridge ignores DELETE
    /// bodies, so `userId` / `projectId` must be in the query string.
    pub async fn unassign(
        &self,
        template_id: &str,
        user_id: Option<&str>,
        project_id: Option<&str>,
    ) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("userId", user_id.map(str::to_owned));
        q.insert("projectId", project_id.map(str::to_owned));
        self.http
            .delete(
                &format!("/api/v1/private/user-templates/{template_id}/assignments"),
                &q,
            )
            .await
    }

    /// List projects the caller is a member of (for use in the share target picker).
    pub async fn share_targets(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get("/api/v1/private/user-templates/share-targets", &q)
            .await
    }
}
